# SPDX-License-Identifier: GPL-3.0-only

diff --git a/dddrv.c b/dddrv.c
--- a/dddrv.c
+++ b/dddrv.c
@@ -37,6 +37,13 @@ Copyright (c) 2023 Jaroslav Hensl <emulator@emulace.cz>
 #include "3d_accel.h"
 #include "vmdahal.h"
 
+#ifdef GSW_DIRECTDRAW
+const static DD32BITDRIVERDATA_t drv_gswhal9x = {
+	"gswhal9x.dll",
+	"DriverInit",
+	0
+};
+#else
 const static DD32BITDRIVERDATA_t drv_vmhal9x = {
 	"vmhal9x.dll",
 	"DriverInit",
@@ -49,6 +56,7 @@ const static DD32BITDRIVERDATA_t drv_vmhal486 = {
 	"DriverInit",
 	0
 };
+#endif
 
 /*
  * pre-declare our HAL fns
@@ -625,6 +633,9 @@ BOOL DDCreateDriverObject(int bReset)
 
 BOOL DDGet32BitDriverName(DD32BITDRIVERDATA_t __far *dd32)
 {
+#ifdef GSW_DIRECTDRAW
+	_fmemcpy(dd32, &drv_gswhal9x, sizeof(drv_gswhal9x));
+#else
 	if(drv_is_p2())
 	{
 		_fmemcpy(dd32, &drv_vmhal9x, sizeof(drv_vmhal9x));
@@ -634,6 +645,7 @@ BOOL DDGet32BitDriverName(DD32BITDRIVERDATA_t __far *dd32)
 	{
 		_fmemcpy(dd32, &drv_vmhal486, sizeof(drv_vmhal486));
 	}
+#endif
 
 	if(DDGetPtr(NULL, &(dd32->dwContext)))
 	{
