# SPDX-License-Identifier: GPL-3.0-only

diff --git a/control.c b/control.c
--- a/control.c
+++ b/control.c
@@ -90,6 +90,7 @@ typedef struct _opengl_icd_t
 
 #define OPENGL_ICD_SIZE sizeof(opengl_icd_t)
 
+#ifndef GSW_NO_OPENGL
 /* software OpenGL (Mesa3D softpipe/llvmpipe) */
 const static opengl_icd_t software_icd = {
   0x2,
@@ -118,6 +119,7 @@ const static opengl_icd_t virgl_icd = {
   0x1,
   "VIRGL"
 };
+#endif
 
 /* VIDEOPARAMETERS */
 static VIDEOPARAMETERS_t tvsetup = {
@@ -198,6 +200,10 @@ LONG WINAPI __loadds Control(LPVOID lpDevice, UINT function,
       switch(function_code)
       {
         case OPENGL_GETINFO:
+#ifdef GSW_NO_OPENGL
+          rc = 0;
+          break;
+#endif
         case OP_FBHDA_SETUP:
         case MOUSETRAILS:
         case VIDEOPARAMETERS:
@@ -218,6 +224,9 @@ LONG WINAPI __loadds Control(LPVOID lpDevice, UINT function,
     } /* QUERYESCSUPPORT */
     case OPENGL_GETINFO: /* input: NULL, output: opengl_icd_t */
     {
+#ifdef GSW_NO_OPENGL
+      rc = 0;
+#else
       if((hda->flags & FB_ACCEL_VMSVGA3D) && ((hda->flags & FB_FORCE_SOFTWARE) == 0))
       {
         _fmemcpy(lpOutput, &vmwsvga_icd, OPENGL_ICD_SIZE);
@@ -239,5 +248,6 @@ LONG WINAPI __loadds Control(LPVOID lpDevice, UINT function,
         }
       }
+#endif
       break;
     } /* OPENGL_GETINFO */
     case DCICOMMAND: /* input ptr DCICMD */
